client/json-rpc/src/async_client/http_client.rs

```diff
@@ -0,0 +1,468 @@
+            return Err(Error::unexpected_invalid_batch_response(rpc_resps));
+        }
+
+        let mut resp_maps: HashMap<usize, JsonRpcResponse> = HashMap::new();
+        for resp in rpc_resps {
+            let id = self.validate(&resp, 0, requests.len() - 1)?;
+            if resp_maps.contains_key(&id) {
+                return Err(Error::unexpected_duplicated_response_id(resp));
+            }
+            resp_maps.insert(id, resp);
+        }
+
+        json_requests
+            .iter()
+            .enumerate()
+            .map(|(i, req)| {
+                resp_maps
+                    .remove(&i)
+                    .ok_or_else(|| Error::unexpected_no_response(req.clone()))
+            })
+            .collect()
+    }
+
+    fn update_state(&self, resp_state: State) -> bool {
+        self.state_manager.update_state(resp_state)
+    }
+
+    fn last_known_state(&self) -> Option<State> {
+        self.state_manager.last_known_state()
+    }
+}
+
+pub struct BroadcastHttpClient {
+    pub http_client: reqwest::Client,
+    pub urls: Vec<reqwest::Url>,
+    pub num_parallel_requests: usize,
+    pub state_manager: StateManager,
+}
+
+impl BroadcastHttpClient {
+    pub fn new<T: reqwest::IntoUrl>(
+        server_urls: Vec<T>,
+        num_parallel_requests: usize,
+    ) -> Result<Self, String> {
+        if num_parallel_requests < 1 {
+            return Err("num_parallel_requests should be >= 1".into());
+        }
+        if num_parallel_requests > server_urls.len() {
+            return Err(format!(
+                "num_parallel_requests({}) should be <= length of server_urls({})",
+                num_parallel_requests,
+                server_urls.len()
+            ));
+        }
+        let server_urls = server_urls
+            .into_iter()
+            .map(|x| x.into_url())
+            .collect::<Result<Vec<_>, reqwest::Error>>()
+            .map_err(|x| format!("{}", x))?;
+
+        let reqwest_client = reqwest::ClientBuilder::new()
+            .use_native_tls()
+            .timeout(defaults::HTTP_REQUEST_TIMEOUT)
+            .build()
+            .map_err(|err| format!("{}", err))?;
+        Ok(Self {
+            http_client: reqwest_client,
+            urls: server_urls,
+            num_parallel_requests,
+            state_manager: StateManager::new(),
+        })
+    }
+    async fn multi_send_json_request<T: for<'de> serde::Deserialize<'de>>(
+        &self,
+        req: serde_json::Value,
+    ) -> Result<Vec<T>, Error> {
+        // Choose `num_parallel_requests` from `self.server_urls` to send requests to
+        let chosen_urls: Vec<reqwest::Url> = self
+            .urls
+            .choose_multiple(&mut rand::thread_rng(), self.num_parallel_requests)
+            .cloned()
+            .collect();
+        let futures = chosen_urls.iter().map(|url| {
+            let url = url.clone();
+            let req = req.clone();
+            let client = self.http_client.clone();
```

Comment:
should we clone? or just use the http client reference? 

The Client holds a connection pool internally, so it is advised that you create one and reuse it. see https://docs.rs/reqwest/0.10.9/reqwest/struct.Client.html

Reply:
request::Client is already a reference (Arc<ClientRef>), so cloning it is safe. Without cloning we run into issues because the reference is moved to `async move ` block.